api/v1alpha1/applicationset_types.go

```diff
@@ -70,21 +70,38 @@ type ApplicationSetTemplateMeta struct {
 	Finalizers  []string          `json:"finalizers,omitempty"`
 }
 
-// ApplicationSetGenerator include list item info
+// ApplicationSetGenerator represents a generator at the top level of an ApplicationSet.
 type ApplicationSetGenerator struct {
 	List                    *ListGenerator        `json:"list,omitempty"`
 	Clusters                *ClusterGenerator     `json:"clusters,omitempty"`
 	Git                     *GitGenerator         `json:"git,omitempty"`
-	Matrix                  *MatrixGenerator      `json:"matrix,omitempty"`
 	SCMProvider             *SCMProviderGenerator `json:"scmProvider,omitempty"`
 	ClusterDecisionResource *DuckTypeGenerator    `json:"clusterDecisionResource,omitempty"`
 	PullRequest             *PullRequestGenerator `json:"pullRequest,omitempty"`
+	Matrix                  *MatrixGenerator      `json:"matrix,omitempty"`
+	Merge                   *MergeGenerator       `json:"merge,omitempty"`
 }
 
-// ApplicationSetBaseGenerator include list item info
-// CRD dosn't support recursive types so we need a different type for the matrix generator
+// ApplicationSetNestedGenerator represents a generator nested within a combination-type generator (MatrixGenerator or
+// MergeGenerator).
+type ApplicationSetNestedGenerator struct {
+	List                    *ListGenerator         `json:"list,omitempty"`
+	Clusters                *ClusterGenerator      `json:"clusters,omitempty"`
+	Git                     *GitGenerator          `json:"git,omitempty"`
+	SCMProvider             *SCMProviderGenerator  `json:"scmProvider,omitempty"`
+	ClusterDecisionResource *DuckTypeGenerator     `json:"clusterDecisionResource,omitempty"`
+	PullRequest             *PullRequestGenerator  `json:"pullRequest,omitempty"`
+	Matrix                  *NestedMatrixGenerator `json:"matrix,omitempty"`
+	Merge                   *NestedMergeGenerator  `json:"merge,omitempty"`
```

Comment:
Im sorry but I don't understand why embedded structs makes it easier to miss nil checks. I believe it is the same.

Reply:
Suppose I have this code:

```go
type Book struct {
	Title string
}

func (b *Book) GetLowerCaseTitle() string {
	return toLowerCase(b.Title)
}
```

Then I decide to refactor to use an embedded struct:

```go
type BookMetadata struct {
	Title string
}

type Book struct {
	BookMetadata
}

func (b *Book) GetLowerCaseTitle() string {
	// this still compiles, because `b.Title` is syntactic sugar for `b.BookMetadata.Title`
	return toLowerCase(b.Title)
}

func createBook() Book {
	return Book{
		BookMetadata: nil,  // this compiles
	}
}
```

After the refactor, I have a bunch of references to now-embedded fields which _look_ immune to nil reference errors, but can still throw nil reference errors.

It would seem that embedding structs isn't like a macro (simply expanding the field names and literally "embedding" them in the containing struct). It's just a bit of syntactic sugar for composition, plus some nice default behavior from the json marshaller.

Reply:
^ that's wrong. After the refactor, `Title` _cannot_ be nil. In my previous attempt, I'd used a pointer to an embedded struct, which could be nil. I'm not sure why I went with the pointer. Time permitting, I'll try with the non-pointer embedded struct.